docs/development/design/rate-limiting.md

```diff
@@ -0,0 +1,37 @@
+# Used Envoy for Rate Limiting
+
+<!-- [Issue #600](https://github.com/redhat-et/issues/600) -->
+
+## Summary
+
+To protect the apiserver service from a misbehaving tenant, the rest APIs should be rate limited. This will allow us to better reason about how to scale the service to a large number of tenants as there will be limit to the ammout of apiserver resources each tenant can consume.
+
+This also translates to a limit of number of devices that a tenant can enroll before he starts to hit rate limits.  There should be a way to increase the limits per tenant.
+
+## Proposal
+
+Use an [envoy](https://www.envoyproxy.io/) as a sidecar proxy and [limitador](https://github.com/Kuadrant/limitador) to enforce the rate limiting policies.
+
+Running the proxy as sidecar has the following benefits:
+
+* More secure since the proxy and apiserver communicate on the loopback interface reducing the possibility of packet snooping.
+* More efficient since the proxy and apiserver are guaranteed to run on the same worker node.
+* Easier to reason about scaling with traffic load since both get scaled up as unit.
+
+Using an Envoy proxy has the following benefits:
+
+* Fast proxy used by many projects and supported by many organizations
+* We could possibly move more functionality to it in the future (like oauth processing)
+* Some K8s platforms are moving to envoy as the Ingress gateway so in the future we may be able to move this functionality into the ingress gateway (avoiding a proxy hop).
+* Aligned with service mesh deployment models.
+
+## Alternatives Considered
+
+Live without rate limiting.  If you have a custom Nexodus service deployment, and don't share it with multiple tenants, you may not need rate limiting.
```

Comment:
This isn't the only alternative though.

The concern that I have is how we communicate the rate-limiting information to the proxy.
It would be ideal if this could be read from the database to support different limits per-tenant.

We could actually embed this logic in the apiserver: https://pkg.go.dev/github.com/gin-gonic/examples/ratelimiter

Or perhaps Limitador has an API that we could drive directly.

Reply:
Maybe we could some info into the JWT and use that to pick the right rate limiting config for the tenant.

Reply:
Given rate limiting is such a common service need, we should try to use something that can be commonly supported across services.

And maybe even more importantly, rate limited requests should be dropped as early as possible in the proxy chain to avoid adding load to the downstream servers.  Like once Gateway API becomes a commonly supported API on k8s, it should be possible to have the ingress gateway use limitador to enforce the rate limits.  Allowing us to drop the request even earlier in the proxy chain.